Add snake demo for Sonata board

AlexJones0 opened the pull request:
I've implemented the Snake demo for the Sonata board to show off the CHERI capabilities.

Currently it is implemented as a fully functional game of Snake, where the snake attempting to exit the boundaries of the display will trigger a CHERI Capability violation, which is used to detect the end of the game. The bottom and right side of the screen trigger bounds violations as we try to access outside of a 2D array representing the game. The top and left side of the screen trigger tag violations due to using a negative index on an array, which is undefined behavior. The snake colliding with itself or winning the game currently just ends the game normally, without any CHERI capability violation.

The demo should be quite customisable - you can change things like the game speed, size and appearance using the constants defined near the top of the file to tweak it as you like.

Would appreciate any feedback regarding where improvements can be made in the code - I'm still quite unfamiliar with many of the features of C++.

HU90m commented on examples/snake/snake.cc:
```diff
@@ -0,0 +1,651 @@
+	{
+		// If an explicit out of bounds access occurs, or bounds are made
+		// invalid by some negative array access, the snake has hit the boundary
+		// of the game and so the game should end.
```
Since we don't check which part of the code this error came from. We should probably state in the comment that we *assume* the cheri violation came from the array access in `check_if_colliding`.

HU90m reviewed (changes requested):
Only did a quick skim, but looks good!

```sh
nix run .#clang-lint fix
```

AlexJones0 commented on examples/snake/xor_shift.cc:
There's no particular reason - I was mirroring the way that [`gpiolib.hh`](examples/all/gpiolib.hh) was being added as a dependency by xmake, to ensure everything was built correctly. On checking I can just remove this compartment and instead use `add_files` in the xmake file so I'll change it to that.

nbdd0121 commented on examples/snake/snake.cc:
```diff
@@ -0,0 +1,651 @@
+	LEFT  = 3
+};
+
+enum class Tile : uint64_t
```
Why uint64_t?

AlexJones0 replied:
We looked into why this was needed - the memory allocater mallocs in chunks of 8 bytes and so certain sizes could cause some issues with the capabilities detecting out of bounds accesses. For now for the purposes of this demo I've just left this as uint64_t to make sure everything works.

nbdd0121 commented on examples/snake/snake.cc:
```diff
@@ -0,0 +1,651 @@
+void size_t_to_str_base10(char *buffer, size_t num)
+{
+	// Parse the digits using repeated remainders mod 10
+	int8_t end_idx = 0;
```
```suggestion
	size_t end_idx = 0;
```
Use `size_t` for array indices (or ptrdiff_t if the index can go negative)

nbdd0121 commented on examples/snake/snake.cc:
```diff
@@ -0,0 +1,651 @@
+class SnakeGame
+{
+	private:
+	bool   isFirstGame = true, lastGameWon = false;
```
Split into two lines

nbdd0121 commented on examples/snake/xor_shift.cc:
```diff
@@ -0,0 +1,27 @@
+	state = seed;
+}
+
+uint32_t __cheri_compartment("xor_shift") xor_rand()
```
If you want to generate random number, use the entropy driver instead.

nbdd0121 commented on libraries/lcd.cc:
```diff
@@ -76,23 +76,33 @@ namespace sonata::lcd::internal
+void __cheri_libcall SonataLcd::clean(Color color)
+{
+	// Clean the display with a rectangle of the given colour
+    size_t w, h;
```
No need to do that, lcd_st7735_get_resolution have these as out parameter.

HU90m replied:
I noticed, but it doesn't hurt to get into the habit of initializing variables and it can save a lot of pain.

HU90m commented on examples/snake/snake.cc:
```diff
@@ -0,0 +1,661 @@
+ *
+ * @return True, always.
+ */
+bool __attribute__((noinline)) return_from_handled_error()
```
```suggestion
bool [[gnu::noinline]] return_from_handled_error()
```

AlexJones0 replied:
```c++
[[gnu::noinline]] bool return_from_handled_error()
``` 
seems to work correctly, thanks!

AlexJones0 commented:
I've fixed up all the raised comments now and changed a couple things based on feedback:
- The snake can now move in any direction when it is only 1 tile long, as this was causing some confusion
- Rather than the original `xor_shift` functionality, the program now uses the superior PRNG entropy driver from https://github.com/lowRISC/cheriot-rtos/tree/ethernet (specifically https://github.com/lowRISC/cheriot-rtos/commit/32cd341c09014b4903c1637552f38e8527c45834).
  - This involved bumping the `cheriot-rtos` submodule to that commit because the relevant PR (https://github.com/microsoft/cheriot-rtos/pull/257) has not yet been merged.
 - Added some extra clarifying comments on specific parts of code that required some more explanation.

From investigation the CI quality check appears to be failing because `nix run .#clang-lint fix` is not equipped to deal with changes to the `cheriot-rtos` submodule. I could just leave the simple randomness using `xor_shift` but I figure it's better to use the entropy driver to avoid the possibility of forgetting to replace this in the future when that is properly merged.

AlexJones0 commented:
> we originally had this working with a PLoad violation but that was causing an issue with the hacky way that we're modifying the PCC and restoring the registers to return back to the program from the error handling, so it was decided to leave that for now

Just in case anybody needs it for future reference this is how I was manually triggering the CHERI PLoad capability violation:

```c++
Capability<Tile *> gameCapability{gameSpace};
gameCapability.permissions() &= PermissionSet::omnipotent().without(Permission::Load);
return gameCapability[nextPosition.y][nextPosition.x] == Tile::SNAKE;
```
